File: Wire-iOS/Sources/UserInterface/Components/Views/UserImageView/UserImageView.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "UserImageView.h"
#import <zmessaging/ZMBareUser+UserSession.h>
#import <ZMCDataModel/ZMBareUser.h>
#import <PureLayout/PureLayout.h>

#import "zmessaging+iOS.h"
#import "ImageCache.h"
#import "UIImage+ImageUtilities.h"
#import "UIImage+ZetaIconsNeue.h"
#import "WAZUIMagicIOS.h"


CGFloat PixelSizeForUserImageSize(UserImageViewSize size);
CGFloat PointSizeForUserImageSize(UserImageViewSize size);

CGFloat PixelSizeForUserImageSize(UserImageViewSize size)
{
    return PointSizeForUserImageSize(size) * [UIScreen mainScreen].scale;
}

CGFloat PointSizeForUserImageSize(UserImageViewSize size)
{
    switch (size) {
        case UserImageViewSizeTiny:
            return 36.0f;
            break;
        case UserImageViewSizeSmall:
            return 56.0f;
            break;
        case UserImageViewSizeNormal:
            return 64.0f;
            break;
        case UserImageViewSizeBig:
            return 320.0f;
            break;
        default:
            break;
    }
    return 0;
}


typedef void (^ImageCacheCompletionBlock)(id , NSString *);



static CIContext *ciContext(void)
{
    static CIContext *context;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        context = [CIContext contextWithOptions:nil];
    });
    return context;
}

@interface UserImageView ()

@property (nonatomic) id userObserverToken;
@property (nonatomic) UIView *indicator;

@end



@implementation UserImageView

- (instancetype)initWithCoder:(NSCoder *)coder
{
    self = [super initWithCoder:coder];
    if (self) {
        [self setupBasicProperties];
    }
    return self;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    self = [super initWithFrame:frame];
    if (self) {
        [self setupBasicProperties];
    }
    return self;
}

- (instancetype)initWithMagicPrefix:(NSString *)magicPrefix
{
    self = [self initWithFrame:CGRectZero];

    if (self) {
        [self setupWithMagicPrefix:magicPrefix];
    }

    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];
    
    self.indicator.layer.cornerRadius = self.indicator.bounds.size.width / 2;
}


- (void)setupBasicProperties
{
    _borderColorMatchesAccentColor = YES;
    _shouldDesaturate = YES;
    _suggestedImageSize = UserImageViewSizeNormal;
    
    [self createIndicator];
    [self createConstraints];
}

- (CGSize)intrinsicContentSize
{
    CGFloat imageSize = PointSizeForUserImageSize(self.suggestedImageSize);
    return CGSizeMake(imageSize, imageSize);
}

- (void)setUser:(id<ZMBareUser, ZMSearchableUser>)user
{    
    _user = user;
    
    if (user != nil && ([user isKindOfClass:[ZMUser class]] || [user isKindOfClass:[ZMSearchUser class]])) {
        self.userObserverToken = [UserChangeInfo addObserver:self forBareUser:user];
    }
    
    self.initials.textColor = UIColor.whiteColor;
    self.initials.text = user.initials.uppercaseString;
    
    self.imageView.contentMode = UIViewContentModeScaleAspectFill;
    
    [self setUserImage:nil];
    [self updateBorderColor];
    [self updateIndicatorColor];
    [self updateUserImage];
}

- (void)createIndicator
{
    self.indicator = [[UIView alloc ] initForAutoLayout];
    self.indicator.backgroundColor = UIColor.redColor;
    self.indicator.hidden = YES;
    
    [self addSubview:self.indicator];
}

- (void)createConstraints
{
    [self.indicator autoPinEdgeToSuperviewEdge:ALEdgeTop];
    [self.indicator autoPinEdgeToSuperviewEdge:ALEdgeRight];
    [self.indicator autoMatchDimension:ALDimensionWidth toDimension:ALDimensionWidth ofView:self.containerView withMultiplier:1.0f / 3.0f];
    [self.indicator autoMatchDimension:ALDimensionHeight toDimension:ALDimensionWidth ofView:self.indicator];
}

- (void)updateIndicatorColor
{
    self.indicator.backgroundColor = [(id)self.user accentColor];
}

- (void)updateBorderColor
{
    if (! self.borderColorMatchesAccentColor) {
        return;
    }
    
    if (self.user.isConnected || self.user.isSelfUser) {
        self.borderColor = [(id)self.user accentColor];
    } else {
        self.borderColor = [UIColor clearColor];
    }
}

- (void)updateUserImage
{
    if (self.suggestedImageSize == UserImageViewSizeBig &&
        self.user.imageMediumData == nil) {
        
        [self.user requestMediumProfileImageInUserSession:[ZMUserSession sharedSession]];
        return;
    }
    
    if (self.suggestedImageSize != UserImageViewSizeBig &&
        self.user.imageSmallProfileData == nil) {
        
        [self.user requestSmallProfileImageInUserSession:[ZMUserSession sharedSession]];
        return;
    }
    
    NSData *imageData = nil;
    NSString *imageCacheKey = nil;
    
    if (self.suggestedImageSize == UserImageViewSizeBig) {
        imageData = self.user.imageMediumData;
        imageCacheKey = self.user.imageMediumIdentifier;
    }
    else {
        imageData = self.user.imageSmallProfileData;
        imageCacheKey = self.user.imageSmallProfileIdentifier;
    }
    BOOL userIsConnected = self.user.isConnected || self.user.isSelfUser;

    if (imageData == nil || imageCacheKey == nil) {
        return;
    }

    // cache key is not changed when user change his own image
    if (self.user.isSelfUser) {
        
        UIImage *image = [UIImage imageFromData:imageData withMaxSize:PixelSizeForUserImageSize(self.suggestedImageSize)];

        [self setUserImage:image];
        return;
    }

    @weakify(self);
    
    ImageCacheCompletionBlock completionBlock = ^void(id image, NSString *cacheKey){
        
        @strongify(self);
        
        NSString *updatedCacheKey = nil;
        
        if (self.suggestedImageSize == UserImageViewSizeBig) {
            updatedCacheKey = self.user.imageMediumIdentifier;
        }
        else {
            updatedCacheKey = self.user.imageSmallProfileIdentifier;
        }
        
        if ([cacheKey isEqualToString:updatedCacheKey]) {
            [self setUserImage:image];
        }
    };
    
    if (userIsConnected || ! self.shouldDesaturate) {
        [[UserImageView sharedFullColorImageCacheForSize:self.suggestedImageSize] imageForData:imageData cacheKey:imageCacheKey creationBlock:^id(NSData *data) {
            
            UIImage *image = [UIImage imageFromData:data withMaxSize:PixelSizeForUserImageSize(self.suggestedImageSize)];
            return image;
            
        } completion:completionBlock];
    }
    else {
        [[UserImageView sharedDesaturatedImageCacheForSize:self.suggestedImageSize] imageForData:imageData cacheKey:imageCacheKey creationBlock:^id(NSData *data) {
            
            UIImage *image = [[UserImageView sharedFullColorImageCacheForSize:self.suggestedImageSize] imageForCacheKey:imageCacheKey];
            
            if (! image) {
                image = [UIImage imageFromData:data withMaxSize:PixelSizeForUserImageSize(self.suggestedImageSize)];
            }
            image = [image desaturatedImageWithContext:ciContext()
                                            saturation:[WAZUIMagic sharedMagic][@"background.image_target_saturation"]];
          
            return image;
            
        } completion:completionBlock];
    }
}

- (void)setUserImage:(UIImage *)userImage
{
    self.initials.hidden = userImage != nil;
    self.imageView.hidden = userImage == nil;
    self.imageView.image = userImage;
    
    if (userImage) {
        self.containerView.backgroundColor = UIColor.clearColor;
    }
    else if (self.user.isConnected || self.user.isSelfUser) {
        self.containerView.backgroundColor = [(id)self.user accentColor];
    }
    else {
        self.containerView.backgroundColor = [UIColor colorWithMagicIdentifier:@"connect.user_not_connected_color"];
    }
}

- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event
{
    [super touchesEnded:touches withEvent:event];
    
    if ([self.delegate respondsToSelector:@selector(userImageViewTouchUpInside:)]) {
        [self.delegate userImageViewTouchUpInside:self];
    }
}

#pragma mark - Indicator

- (void)setIndicatorEnabled:(BOOL)indicatorEnabled
{
    _indicatorEnabled = indicatorEnabled;
    
    self.indicator.hidden = ! indicatorEnabled;
}

#pragma mark - ZMUserObserver

- (void)userDidChange:(UserChangeInfo *)change
{
    if (self.suggestedImageSize == UserImageViewSizeBig) {
        if (change.imageMediumDataChanged || change.connectionStateChanged) {
            [self updateBorderColor];
            [self updateUserImage];
        }
    }
    else {
        if (change.imageSmallProfileDataChanged || change.connectionStateChanged) {
            [self updateBorderColor];
            [self updateUserImage];
        }
    }
    
    if (change.accentColorValueChanged) {
        [self updateBorderColor];
        [self updateIndicatorColor];
    }
}

#pragma mark - Class Methods

+ (ImageCache *)sharedFullColorImageCacheForSize:(UserImageViewSize)size
{
    static NSArray *fullColorImageCaches;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSMutableArray *fullColorImageCachesMutable = [NSMutableArray arrayWithCapacity:UserImageViewSizeLast - UserImageViewSizeFirst];
        
        for (UserImageViewSize s = UserImageViewSizeFirst; s <= UserImageViewSizeLast; s++) {            
            ImageCache *fullColorImageCache = [[ImageCache alloc] initWithName:[NSString stringWithFormat:@"_UserImageView.fullColorImageCache_%d", (int)s]];
            fullColorImageCache.maxConcurrentOperationCount = 4;
            fullColorImageCache.countLimit = 100;
            fullColorImageCache.totalCostLimit = 1024 * 1024 * 10;
            fullColorImageCache.qualityOfService = NSQualityOfServiceUtility;
            [fullColorImageCachesMutable addObject:fullColorImageCache];
        }
        
        fullColorImageCaches = fullColorImageCachesMutable;
    });
    return fullColorImageCaches[size];
}

+ (ImageCache *)sharedDesaturatedImageCacheForSize:(UserImageViewSize)size
{
    static NSArray *desaturatedImageCaches;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        NSMutableArray *desaturatedImageCachesMutable = [NSMutableArray arrayWithCapacity:UserImageViewSizeLast - UserImageViewSizeFirst];
        
        for (UserImageViewSize s = UserImageViewSizeFirst; s <= UserImageViewSizeLast; s++) {            
            ImageCache *desaturatedImageCache = [[ImageCache alloc] initWithName:[NSString stringWithFormat:@"_UserImageView.desaturatedImageCache_%d", (int)s]];
            desaturatedImageCache.maxConcurrentOperationCount = 4;
            desaturatedImageCache.countLimit = 100;
            desaturatedImageCache.totalCostLimit = 1024 * 1024 * 10;
            desaturatedImageCache.qualityOfService = NSQualityOfServiceUtility;
            [desaturatedImageCachesMutable addObject:desaturatedImageCache];
        }
        
        desaturatedImageCaches = desaturatedImageCachesMutable;
    });
    return desaturatedImageCaches[size];
}

@end

@implementation UserImageView (Magic)

- (void)setupWithMagicPrefix:(NSString *)prefix
{
    if (0 < [prefix length]) {
        self.borderWidth = [WAZUIMagic cgFloatForIdentifier:[self magicPathForKey:@"stroke_width" withPrefix:prefix]];
        self.initials.font = [UIFont fontWithMagicIdentifier:[self magicPathForKey:@"user_initials_font" withPrefix:prefix]];
        self.initials.textColor = [UIColor colorWithMagicIdentifier:[self magicPathForKey:@"user_initials_font_color" withPrefix:prefix]];
    }
}

- (NSString *)magicPathForKey:(NSString *)key withPrefix:(NSString *)prefix
{
    return [NSString stringWithFormat:@"%@.%@", prefix, key];
}

@end

File: wire-ios/Wire-iOS/Sources/UserInterface/Components/UserImageView.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import WireSyncEngine

/// A view that displays the avatar for a remote user.
class UserImageView: AvatarImageView, UserObserving {
    // MARK: Lifecycle

    // MARK: - Initialization

    override init(frame: CGRect) {
        self.size = .small
        super.init(frame: .zero)
        configureSubviews()
        configureConstraints()
    }

    init(size: Size = .small) {
        self.size = size
        super.init(frame: .zero)
        configureSubviews()
        configureConstraints()
    }

    @available(*, unavailable)
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) is not supported")
    }

    // MARK: Internal

    // MARK: - Nested Types

    /// The different sizes for the avatar image.
    enum Size: Int {
        case tiny = 16
        case badge = 24
        case small = 32
        case normal = 64
        case big = 320
    }

    // MARK: - Interface Properties

    /// The size of the avatar.
    var size: Size {
        didSet { updateUserImage() }
    }

    /// Whether the image should be desaturated, e.g. for unconnected users.
    var shouldDesaturate = true {
        didSet { updateUserImage() }
    }

    /// Whether the badge indicator is enabled.
    var indicatorEnabled = false {
        didSet { badgeIndicator.isHidden = !indicatorEnabled }
    }

    // MARK: - Remote User

    /// The user session to use to download images.
    var userSession: UserSession? {
        didSet { updateUser() }
    }

    /// The user to display the avatar of.
    var user: UserType? {
        didSet { updateUser() }
    }

    override var intrinsicContentSize: CGSize {
        .init(width: size.rawValue, height: size.rawValue)
    }

    // MARK: - Changing the Content

    /// Sets the avatar for the user with an optional animation.
    /// - parameter avatar: The avatar of the user.
    /// - parameter user: The currently displayed user.
    /// - parameter animated: Whether to animate the change.
    func setAvatar(_ avatar: Avatar, user: UserType, animated: Bool) {
        let updateBlock = {
            self.avatar = avatar
            self.container.backgroundColor = self.containerBackgroundColor(for: user)
        }

        if animated, !ProcessInfo.processInfo.isRunningTests {
            UIView.transition(
                with: self,
                duration: 0.15,
                options: .transitionCrossDissolve,
                animations: updateBlock,
                completion: nil
            )
        } else {
            updateBlock()
        }
    }

    // MARK: - Updates

    func userDidChange(_ changeInfo: UserChangeInfo) {
        // Check for potential image changes
        if size == .big {
            if changeInfo.imageMediumDataChanged || changeInfo.connectionStateChanged {
                updateUserImage()
            }
        } else {
            if changeInfo.imageSmallProfileDataChanged || changeInfo.connectionStateChanged || changeInfo.teamsChanged {
                updateUserImage()
            }
        }

        // Change for accent color changes
        if changeInfo.accentColorValueChanged {
            updateIndicatorColor()
        }
    }

    /// Called when the user or user session changes.
    func updateUser() {
        guard let user, let initials = user.initials else {
            return
        }

        let defaultAvatar: Avatar = initials.isEmpty ? .init() : .text(initials.localizedUppercase)
        setAvatar(defaultAvatar, user: user, animated: false)
        if !ProcessInfo.processInfo.isRunningTests,
           let userSession = userSession as? ZMUserSession {
            userObserverToken = UserChangeInfo.add(observer: self, for: user, in: userSession)
        }

        updateForServiceUserIfNeeded(user)
        updateIndicatorColor()
        updateUserImage()
    }

    // MARK: Fileprivate

    /// Updates the image for the user.
    fileprivate func updateUserImage() {
        guard
            let user,
            let userSession
        else {
            return
        }

        var desaturate = false
        if shouldDesaturate {
            desaturate = !user.isConnected && !user.isSelfUser && !user.isTeamMember && !user.isServiceUser
        }

        user.fetchProfileImage(
            session: userSession,
            imageCache: UIImage.defaultUserImageCache,
            sizeLimit: size.rawValue,
            isDesaturated: desaturate
        ) { [weak self] image, cacheHit in
            // Don't set image if nil or if user has changed during fetch
            guard let image, user.isEqual(self?.user) else { return }
            self?.setAvatar(.image(image), user: user, animated: !cacheHit)
        }
    }

    // MARK: Private

    private let badgeIndicator = RoundedView()

    private var userObserverToken: NSObjectProtocol?

    private func configureSubviews() {
        accessibilityElementsHidden = true

        badgeIndicator.backgroundColor = .red
        badgeIndicator.isHidden = true
        badgeIndicator.shape = .circle
        addSubview(badgeIndicator)
    }

    private func configureConstraints() {
        badgeIndicator.translatesAutoresizingMaskIntoConstraints = false

        setContentHuggingPriority(.required, for: .vertical)
        setContentHuggingPriority(.required, for: .horizontal)

        NSLayoutConstraint.activate([
            badgeIndicator.topAnchor.constraint(equalTo: topAnchor),
            badgeIndicator.trailingAnchor.constraint(equalTo: trailingAnchor),
            badgeIndicator.heightAnchor.constraint(equalTo: badgeIndicator.widthAnchor),
            badgeIndicator.widthAnchor.constraint(equalTo: widthAnchor, multiplier: 1 / 3),
        ])
    }

    // MARK: - Interface

    /// Returns the appropriate border width for the user.
    private func borderWidth(for user: UserType) -> CGFloat {
        user.isServiceUser ? 0.5 : 0
    }

    /// Returns the appropriate border color for the user.
    private func borderColor(for user: UserType) -> CGColor? {
        user.isServiceUser ? UIColor.black.withAlphaComponent(0.08).cgColor : nil
    }

    /// Returns the placeholder background color for the user.
    private func containerBackgroundColor(for user: UserType) -> UIColor {
        switch avatar {
        case .image:
            user.isServiceUser ? .white : .clear
        case .text:
            if user.isConnected || user.isSelfUser || user.isTeamMember || user.isWirelessUser {
                user.accentColor
            } else {
                .init(white: 0.8, alpha: 1)
            }
        }
    }

    /// Returns the appropriate avatar shape for the user.
    private func shape(for user: UserType) -> AvatarImageView.Shape {
        user.isServiceUser ? .relative : .circle
    }

    /// Updates the color of the badge indicator.
    private func updateIndicatorColor() {
        badgeIndicator.backgroundColor = user?.accentColor
    }

    /// Updates the interface to reflect if the user is a service user or not.
    private func updateForServiceUserIfNeeded(_ user: UserType) {
        let oldValue = shape
        shape = shape(for: user)
        if oldValue != shape {
            container.layer.borderColor = borderColor(for: user)
            container.layer.borderWidth = borderWidth(for: user)
            container.backgroundColor = containerBackgroundColor(for: user)
        }
    }
}
